dev-tool/src/list_net/data_source.rs

```diff
@@ -0,0 +1,567 @@
+        bincode::DefaultOptions::new()
+            .serialize_into(writer, self)
+            .map_err(Into::into)
+    }
+
+    /// Deserialize a instance from given file.
+    #[allow(dead_code)] //FIXME is used by training (added in part 3 of this PR)
+    pub(crate) fn deserialize_from_file(file: impl AsRef<Path>) -> Result<Self, Error> {
+        Self::deserialize_from(BufReader::new(File::open(file)?))
+    }
+
+    /// Deserialize a instance from given reader, preferably pass in a buffered reader.
+    #[allow(dead_code)] //FIXME is used by training (added in part 3 of this PR)
+    fn deserialize_from(reader: impl Read) -> Result<Self, Error> {
+        let self_: Self = bincode::DefaultOptions::new().deserialize_from(reader)?;
+        debug!("Loaded {} samples.", self_.data.len());
+        Ok(self_)
+    }
+
+    fn load_sample_helper(&self, id: DataId) -> Result<Sample, StorageError> {
+        let raw = &self.data[id];
+
+        // len == nr_document * nr_features + nr_documents * 1
+        let nr_documents = raw.len() / (ListNet::INPUT_NR_FEATURES + 1);
+        let start_of_target_prob_dist = nr_documents * ListNet::INPUT_NR_FEATURES;
+        debug_assert_eq!(start_of_target_prob_dist + nr_documents, raw.len());
+
+        let inputs = ArrayView::from_shape(
+            (nr_documents, ListNet::INPUT_NR_FEATURES),
+            &raw[..start_of_target_prob_dist],
+        )
+        .map_err(|_| StorageError::BrokenInvariants { at_index: id })?;
+        let target_prob_dist =
+            ArrayView::from_shape((nr_documents,), &raw[start_of_target_prob_dist..])
+                .map_err(|_| StorageError::BrokenInvariants { at_index: id })?;
+
+        Ok(Sample {
+            inputs,
+            target_prob_dist,
+        })
+    }
+
+    /// Add a new sample.
+    ///
+    /// # Panics
+    ///
+    /// - If the number of documents in `inputs` doesn't match the number of probabilities in
+    /// `target_prob_dist`.
+    /// - If the number of features per document is not equal to `[ListNet::INPUT_NR_FEATURES]`.
+    pub(crate) fn add_sample(
+        &mut self,
+        inputs: ArrayView2<f32>,
+        target_prob_dist: ArrayView1<f32>,
+    ) -> Result<(), Error> {
+        if inputs.shape() != [target_prob_dist.len(), ListNet::INPUT_NR_FEATURES] {
+            bail!("Sample with bad array shapes. Expected shapes [{nr_docs}, {nr_feats}] & [{nr_docs}] but got shapes {inputs_shape:?} & {prob_dist_shape:?}",
+                nr_docs=inputs.shape()[0],
+                nr_feats=ListNet::INPUT_NR_FEATURES,
+                inputs_shape=inputs.shape(),
+                prob_dist_shape=target_prob_dist.shape(),
+            );
+        }
+        let mut data = Vec::with_capacity(inputs.len() + target_prob_dist.len());
+        extend_vec_with_ndarray(&mut data, inputs);
+        extend_vec_with_ndarray(&mut data, target_prob_dist);
+        self.data.push(data);
```

Comment:
perhaps you could call `data` on l336 something else e.g. `datum` so that e.g.
```rust
self.data.push(datum);
```
looks a bit less confusing